### ISSUE TITLE
What if `SET` took an array of objects

### ISSUE THREAD

**Issue by 96LawDawg** [ArnoldSmith86/virtualtabletop#1168]
I just had a crazy idea.  Not even sure if "array of objects" is the right phrase.  I'm thinking that `SET` could continue to work as is for simple cases, or have a different syntax like the new css for complicated cases.  So in addition to what we have, you add a "multiset" option that looks like this:
{
      "func": "SET",
      "multiset": {
          "widget1": {
             "property1": "${whatever}",
             "property2": 7,
             "property3": true
            },
       "widget2": {
             "property1": "Text",
             "property2": "${PROPERTY x}"
            }
       }
 },

**Comment by robartsd** [comment]
I think setting different values for different widgets should remain separate `SET` blocks; however, I would be fine with setting multiple properties in the widget(s) in a single `SET` block.

**Comment by rogerl50** [comment]
I just needed that today as well!

**Comment by 96LawDawg** [comment]
I was just about to say that.  Roger, I think anonymous collections were your idea.  This already works:

```JSON
{
      "func": "SET",
      "collection": [
        "abc1",
        "abc2",
        "abc3"
      ],
      "property": "y",
      "value": 0
    }
```

**Comment by rogerl50** [comment]
> > I just needed that today as well! I would also like the widget name to be an array so you could set the same property in multiple widgets at once.
> > As an alternative (or in addition) how about allowing SET to accept a collection?
> 
> SET already uses collections (including anonymous collections).

Right, I knew that :smiley:. I was typing from memory - what I ran across today was the fact that `CALL` does not take a collection. That would be handy for calling the same routine in multiple similar objects (like dice).

**Comment by robartsd** [comment]
> > > I just needed that today as well! I would also like the widget name to be an array so you could set the same property in multiple widgets at once.
> > > As an alternative (or in addition) how about allowing SET to accept a collection?
> > 
> > 
> > SET already uses collections (including anonymous collections).
> 
> Right, I knew that 😃. I was typing from (my lousy) memory - what I ran across today was the fact that `CALL` does not take a collection. That would be handy for calling the same routine in multiple similar objects (like dice).

I recommend FOREACH for that.
```json
{
  "func": "FOREACH",
  "collection": collection,
  "loopRoutine": [
    {
      "func": "CALL",
      "widget": "${widgetID}",
      "routine": routineName
    }
  ]
}
```
Makes it clear that the routines are not called in parallel (they can't be).

**Comment by ArnoldSmith86** [comment]
Been wanting to add this for a while (every time I `SET` `x` and `y` of something).

I'm not sure if we need multiple widget targets like in the initial example but why not? I would have used that several times.

**Comment by robartsd** [comment]
> I'm not sure if we need multiple widget targets like in the initial example but why not?

Mostly I was thinking the syntax could be troublesome and not significantly more readable or concise than a series of SET blocks. Since a widget property can take any value, we can't simply overload like we did with the `css` property. I suppose the `assign` parameter could take an object (key value pairs are assigned to all widgets in collection) or array of objects (first widget in collecton gets assigned first element of assign array, and so forth) - that could be useful in some circumstances.

**Comment by ArnoldSmith86** [comment]
> Since a widget property can take any value, we can't simply overload like we did with the `css` property.

Didn't think of that.

**Comment by RaphaelAlvez** [comment]
wouldn't this be fairly solved with the compact version of routines that we already talked about? just something like:
SET x = 5
SET Y - 80

**Comment by rogerl50** [comment]
See also #579. These two should be merged.

**Comment by flackr** [comment]
Thanks for marging, I didn't catch this. From #1404 I proposed an alternative where the properties of SET could take list values and auto-extend similar to one of the ways CSS treats different list lengths for some properties. e.g.

```
{
  "func": "SET",
  "property": ["x", "y", "parent", "owner"],
  "value": [100, 200, "container", "player"]
  }
}
```

Auto-extending the last value. This could be useful for different relations, e.g.
```
{
  "func": "SET",
  "property": ["x", "y", "parent", "owner"],
  "relation": ["+", "+", "="],
  "value": [100, 200, "container", "player"]
  }
}
```
